divi/src/wallet.h

```diff
@@ -178,6 +178,11 @@ class CWallet : public CCryptoKeyStore, public NotificationInterface, public I_K
     //it was public bool SelectCoins(int64_t nTargetValue, std::set<std::pair<const CWalletTx*,unsigned int> >& setCoinsRet, int64_t& nValueRet, const CCoinControl *coinControl = NULL, AvailableCoinsType coin_type=ALL_SPENDABLE_COINS, bool useIX = true) const;
     void DeriveNewChildKey(const CKeyMetadata& metadata, CKey& secretRet, uint32_t nAccountIndex, bool fInternal /*= false*/);
 
+    /** Returns true if the wallet should allow spending of unconfirmed change.
+     *  This is mostly determined by allowSpendingZeroConfirmationOutputs,
+     *  but is forced to off around the segwit-light fork.  */
+    bool AllowSpendingZeroConfirmationChange() const;
```

Comment:
Wouldn't it be simpler to do the upgrade process in two separate upgrades?
(1) You change the GetHash function to return the legacy hash if the transaction version is some specific value (the present default) and to return the bare txid for the upgraded version number.
(2) You upgrade the network as a soft fork by not including the updated version number directly.
(3) Once everybody has made the first update, the second update is a softfork relative to the first one.

What are the limitations to doing this?

Reply:
That is an interesting suggestion.  I don't see any reason why it wouldn't work.  However, I believe it would basically just make things more complex without much benefit:
- it would still be a hard fork (one of them, namely step 1), so no benefit in this respect
- it would be two forks instead of one that need to be scheduled, people notified, and hoping that they update in time, which I think rather increases than decreases the overall risk
- instead of a rather simple rule change "in one place" (concerning consensus at least), namely what UTXOs get created for a particular transaction, we now have a more complex dependency between the transaction version, things the version implies, and consensus rules about what versions are allowed

I can see one benefit of this change, and that is fixing the spending of unconfirmed outputs in a time-window "around" fork activation (which is probably why you brought up this suggestion here).  However, I would much rather have some temporary logic to avoid issues as in this PR (which, notably, also does not affect the core consensus at all) than more complicated fork logic in general that we need to keep for the future.

Reply:
Some clarification on item (1), you wouldn't update the version number in step (1) so that it remains a soft-fork. You'd update it at a later stage after you've ensured enough people have upgraded. Blocks with the hard fork active would only accept newer-version transactions perhaps? Sounds messier I agree. I'm not entirely certain about the best way to go about it.

Reply:
I've thought about this a bit more.  In fact, I think it needs three forks, which at least conceptually are separate (but of course they could under some circumstances be part of the same code update, just scheduled at different times):

1. A soft fork that disallows a certain transaction version completely (or all versions larger than current default).
2. A hard fork that changes what UTXOs get generated for transactions with a new version (which aren't valid on the network at this stage at all, so this change is rather easy to do after the network has updated and activated step 1 successfully).
3. Another hard fork that actually allows transactions of the new version again.

Step 1 is relatively uncritical as a straight-forward soft fork.  But once step 1 is activated on the network, steps 2 & 3 could be done in a follow-up with perhaps a bit less risk and a combined update.  They remain a hard fork, though.

I think this process could work, and might be able to reduce the upgrade risk.  But I'm still not fully convinced it is the best way forward; the only real benefit would be removing that temporary restriction of spending unconfirmed outputs around the activation of the current PR.  And the version-number-dependent logic in itself is then something a bit more complex that remains forever.